ramips: add support for Motorola MWR03

DragonBluep opened the pull request:
Specifications:
* SOC: MT7628AN + MT7612E
* ROM: 8 MiB Flash
* RAM: 64 MiB DDR2
* WAN: 10/100M *1
* LAN: 10/100M *3
* Button: Reset *1
* LEDs: orange *1, white *1
* Antennas: 2.4 GHz *2 + 5 GHz *2
* TTL Baudrate: 57600
* TFTP Upgrade: IP: 192.168.51.1, Server: 192.168.51.100

Installation (TFTP + U-Boot):
 * Connect device with a TTL cable and open a serial session by PuTTY.
 * Press "2" when booting to select "Load system code then write Flash via TFTP".
 * Configure the IP of local host server.
 * Upload firmware by tftpd64, it will boot when write instruction is executed.

DragonBluep commented:
OpenWrt Kernel Log:

[    0.000000] Linux version 5.4.110 (dragon@debian) (gcc version 8.4.0 (OpenWrt GCC 8.4.0 r16485-cae3e37383)) #0 Fri Apr 16 04:45:38 2021
[    0.000000] Board has DDR2
[    0.000000] Analog PMU set to hw control
[    0.000000] Digital PMU set to hw control
[    0.000000] SoC Type: MediaTek MT7628AN ver:1 eco:2
[    0.000000] printk: bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 00019655 (MIPS 24KEc)
[    0.000000] MIPS: machine is Motorola MWR03
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, PIPT, no aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000]   Normal zone: 144 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16240
[    0.000000] Kernel command line: console=ttyS0,57600 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes, linear)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes, linear)
[    0.000000] Writing ErrCtl register=0003d110
[    0.000000] Readback ErrCtl register=0003d110
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory: 57452K/65536K available (4737K kernel code, 202K rwdata, 1048K rodata, 1172K init, 197K bss, 8084K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 256
[    0.000000] intc: using register map from devicetree
[    0.000000] random: get_random_bytes called from start_kernel+0x308/0x4fc with crng_init=0
[    0.000000] CPU Clock: 580MHz
[    0.000000] timer_probe: no matching timers found
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6590553264 ns
[    0.000010] sched_clock: 32 bits at 290MHz, resolution 3ns, wraps every 7405115902ns
[    0.015375] Calibrating delay loop... 385.02 BogoMIPS (lpj=770048)
[    0.059450] pid_max: default: 32768 minimum: 301
[    0.068781] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.083117] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.104765] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 7645041785100000 ns
[    0.123972] futex hash table entries: 256 (order: -1, 3072 bytes, linear)
[    0.137488] pinctrl core: initialized pinctrl subsystem
[    0.148875] NET: Registered protocol family 16
[    0.266926] mt7620-pci 10140000.pcie: Port 0 N_FTS = 1b105000
[    0.427763] PCI host bridge /pcie@10140000 ranges:
[    0.437108]  MEM 0x0000000020000000..0x000000002fffffff
[    0.447418]   IO 0x0000000010160000..0x000000001016ffff
[    0.480831] PCI host bridge to bus 0000:00
[    0.488881] pci_bus 0000:00: root bus resource [mem 0x20000000-0x2fffffff]
[    0.502388] pci_bus 0000:00: root bus resource [io  0xffffffff]
[    0.514084] pci_bus 0000:00: root bus resource [??? 0x00000000 flags 0x0]
[    0.527499] pci_bus 0000:00: No busn resource found for root bus, will use [bus 00-ff]
[    0.543213] pci 0000:00:00.0: [14c3:0801] type 01 class 0x060400
[    0.555057] pci 0000:00:00.0: reg 0x10: [mem 0x00000000-0x7fffffff]
[    0.567412] pci 0000:00:00.0: reg 0x14: [mem 0x20200000-0x2020ffff]
[    0.579874] pci 0000:00:00.0: supports D1
[    0.587698] pci 0000:00:00.0: PME# supported from D0 D1 D3hot
[    0.600768] pci 0000:01:00.0: [14c3:7612] type 00 class 0x028000
[    0.612621] pci 0000:01:00.0: reg 0x10: [mem 0x00000000-0x000fffff 64bit]
[    0.625987] pci 0000:01:00.0: reg 0x30: [mem 0x00000000-0x0000ffff pref]
[    0.639301] pci 0000:01:00.0: PME# supported from D0 D3hot D3cold
[    0.652718] pci_bus 0000:01: busn_res: [bus 01-ff] end is updated to 01
[    0.665707] pci_bus 0000:00: busn_res: [bus 00-ff] end is updated to 01
[    0.678779] pci 0000:00:00.0: BAR 0: no space for [mem size 0x80000000]
[    0.691817] pci 0000:00:00.0: BAR 0: failed to assign [mem size 0x80000000]
[    0.705580] pci 0000:00:00.0: BAR 8: assigned [mem 0x20000000-0x200fffff]
[    0.718999] pci 0000:00:00.0: BAR 9: assigned [mem 0x20100000-0x201fffff pref]
[    0.733273] pci 0000:00:00.0: BAR 1: assigned [mem 0x20200000-0x2020ffff]
[    0.746701] pci 0000:01:00.0: BAR 0: assigned [mem 0x20000000-0x200fffff 64bit]
[    0.761156] pci 0000:01:00.0: BAR 6: assigned [mem 0x20100000-0x2010ffff pref]
[    0.775410] pci 0000:00:00.0: PCI bridge to [bus 01]
[    0.785219] pci 0000:00:00.0:   bridge window [mem 0x20000000-0x200fffff]
[    0.798631] pci 0000:00:00.0:   bridge window [mem 0x20100000-0x201fffff pref]
[    0.813019] workqueue: max_active 576 requested for napi_workq is out of range, clamping between 1 and 512
[    0.836119] clocksource: Switched to clocksource MIPS
[    0.847425] NET: Registered protocol family 2
[    0.856877] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes, linear)
[    0.873357] TCP established hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.888434] TCP bind hash table entries: 1024 (order: 0, 4096 bytes, linear)
[    0.902358] TCP: Hash tables configured (established 1024 bind 1024)
[    0.915061] UDP hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.927906] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.941999] NET: Registered protocol family 1
[    0.950590] PCI: CLS 0 bytes, default 32
[    0.962951] workingset: timestamp_bits=14 max_order=14 bucket_order=0
[    0.984337] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.995766] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    1.032547] mt7621_gpio 10000600.gpio: registering 32 gpios
[    1.043809] mt7621_gpio 10000600.gpio: registering 32 gpios
[    1.055026] mt7621_gpio 10000600.gpio: registering 32 gpios
[    1.066271] Serial: 8250/16550 driver, 3 ports, IRQ sharing disabled
[    1.079995] printk: console [ttyS0] disabled
[    1.088435] 10000c00.uartlite: ttyS0 at MMIO 0x10000c00 (irq = 28, base_baud = 2500000) is a 16550A
[    1.106239] printk: console [ttyS0] enabled
[    1.122714] printk: bootconsole [early0] disabled
[    1.142593] spi-mt7621 10000b00.spi: sys_freq: 193333333
[    1.168731] spi-nor spi0.0: s25fl064k (8192 Kbytes)
[    1.178466] 4 fixed-partitions partitions found on MTD device spi0.0
[    1.191053] Creating 4 MTD partitions on "spi0.0":
[    1.200555] 0x000000000000-0x000000030000 : "bootloader"
[    1.212247] 0x000000030000-0x000000040000 : "config"
[    1.223119] 0x000000040000-0x000000050000 : "factory"
[    1.234216] 0x000000050000-0x000000800000 : "firmware"
[    1.248702] 2 uimage-fw partitions found on MTD device firmware
[    1.260493] Creating 2 MTD partitions on "firmware":
[    1.270342] 0x000000000000-0x0000001e739a : "kernel"
[    1.281276] 0x0000001e739a-0x0000007b0000 : "rootfs"
[    1.292206] mtd: device 5 (rootfs) set to be root filesystem
[    1.305170] 1 squashfs-split partitions found on MTD device rootfs
[    1.317500] 0x0000004f0000-0x0000007b0000 : "rootfs_data"
[    1.330053] libphy: Fixed MDIO Bus: probed
[    1.349759] rt3050-esw 10110000.esw: link changed 0x00
[    1.362004] mtk_soc_eth 10100000.ethernet eth0: mediatek frame engine at 0xb0100000, irq 5
[    1.380475] NET: Registered protocol family 10
[    1.393529] Segment Routing with IPv6
[    1.400997] NET: Registered protocol family 17
[    1.409875] 8021q: 802.1Q VLAN Support v1.8
[    1.425309] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    1.445923] Freeing unused kernel memory: 1172K
[    1.454919] This architecture does not have kernel memory protection.
[    1.467673] Run /sbin/init as init process
[    1.700131] random: fast init done
[    2.089031] init: Console is alive
[    2.096186] init: - watchdog -
[    2.848785] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    2.971247] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    2.996225] init: - preinit -
[    4.423901] random: jshn: uninitialized urandom read (4 bytes read)
[    4.603376] random: jshn: uninitialized urandom read (4 bytes read)
[    4.834022] random: jshn: uninitialized urandom read (4 bytes read)
[    5.334780] rt3050-esw 10110000.esw: link changed 0x00
[    5.560539] urandom_read: 2 callbacks suppressed
[    5.560548] random: procd: uninitialized urandom read (4 bytes read)
[    9.461457] rt3050-esw 10110000.esw: link changed 0x02
[    9.738704] mount_root: jffs2 not ready yet, using temporary tmpfs overlay
[    9.756994] urandom-seed: Seed file not found (/etc/urandom.seed)
[    9.883333] procd: - early -
[    9.889347] procd: - watchdog -
[   10.467249] procd: - watchdog -
[   10.474248] procd: - ubus -
[   10.559590] random: ubusd: uninitialized urandom read (4 bytes read)
[   10.574861] random: ubusd: uninitialized urandom read (4 bytes read)
[   10.591854] random: ubusd: uninitialized urandom read (4 bytes read)
[   10.611116] procd: - init -
[   11.772093] kmodloader: loading kernel modules from /etc/modules.d/*
[   11.833472] Loading modules backported from Linux version v5.10.16-0-gde53befa79cf
[   11.848562] Backport generated by backports.git v5.10.16-1-0-g21d2a1d2
[   11.940677] xt_time: kernel timezone is -0000
[   12.184610] mt76_wmac 10300000.wmac: ASIC revision: 76280001
[   12.325607] urngd: v1.0.2 started.
[   12.526099] random: crng init done
[   12.532871] random: 1 urandom warning(s) missed due to ratelimiting
[   13.246528] mt76_wmac 10300000.wmac: Firmware Version: 20151201
[   13.258337] mt76_wmac 10300000.wmac: Build Time: 20151201183641
[   13.316157] mt76_wmac 10300000.wmac: firmware init done
[   13.496511] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   13.604656] mt76x2e 0000:01:00.0: card - bus=0x1, slot = 0x0 irq=4
[   13.617237] mt76x2e 0000:01:00.0: ASIC revision: 76120044
[   14.802597] mt76x2e 0000:01:00.0: ROM patch build: 20141115060606a
[   14.820866] mt76x2e 0000:01:00.0: Firmware Version: 0.0.00
[   14.831769] mt76x2e 0000:01:00.0: Build: 1
[   14.839886] mt76x2e 0000:01:00.0: Build Time: 201507311614____
[   14.872258] mt76x2e 0000:01:00.0: Firmware running!
[   14.990260] ieee80211 phy1: Selected rate control algorithm 'minstrel_ht'
[   15.022579] PPP generic driver version 2.4.2
[   15.034624] NET: Registered protocol family 24
[   15.062503] kmodloader: done loading kernel modules from /etc/modules.d/*
[   32.814654] rt3050-esw 10110000.esw: link changed 0x00
[   36.894357] rt3050-esw 10110000.esw: link changed 0x02
[   38.912330] jffs2_scan_eraseblock(): End of filesystem marker found at 0x0
[   38.952174] jffs2_build_filesystem(): unlocking the mtd device... 
[   38.952245] done.
[   38.968312] jffs2_build_filesystem(): erasing all blocks after the end marker... 
[   40.224714] br-lan: port 1(eth0.1) entered blocking state
[   40.250291] br-lan: port 1(eth0.1) entered disabled state
[   40.261420] device eth0.1 entered promiscuous mode
[   40.270972] device eth0 entered promiscuous mode
[   40.460875] br-lan: port 1(eth0.1) entered blocking state
[   40.471615] br-lan: port 1(eth0.1) entered forwarding state
[   41.224670] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   48.147793] done.
[   48.151714] jffs2: notice: (1881) jffs2_build_xattr_subsystem: complete building xattr subsystem, 0 of xdatum (0 unchecked, 0 orphan) and 0 of xref (0 dead, 0 orphan) found.
[   48.522783] overlayfs: upper fs does not support tmpfile.

DragonBluep commented:
Add some device picture.

![IMG_20210417_001420](https://user-images.githubusercontent.com/70847398/115053785-19711080-9f12-11eb-82be-31abe7472777.jpg)
![IMG_20210413_234155](https://user-images.githubusercontent.com/70847398/115053855-2a218680-9f12-11eb-93da-1563f2a835f5.jpg)

adschm commented on target/linux/ramips/dts/mt7628an_motorola_mwr03.dts:
```diff
@@ -0,0 +1,140 @@
+};
+
+&wmac {
+	status = "okay";
```
You need to add mtd-eeprom here due to https://github.com/openwrt/openwrt/commit/ad6b0770493e0fd332db9a7fe09f9214342c578b

DragonBluep replied:
Fine, added.

adschm commented:
Looks good so far.

MAC addresses are a bit unexpected (with -1 and +2 for the offsets). Are there addresses in 0x28, 0x2e, 0xe000, 0xe006?

Please also add an overview to the commit message:
https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses

Any chance to have a flashing method without serial access?

DragonBluep committed:
ramips: add support for Motorola MWR03

Specifications:
* SOC: MT7628AN + MT7612E
* ROM: 8 MiB Flash
* RAM: 64 MiB DDR2
* WAN: 10/100M *1
* LAN: 10/100M *3
* Button: Reset *1
* LEDs: orange *1, white *1
* Antennas: 2.4 GHz *2 + 5 GHz *2
* TTL Baudrate: 57600
* TFTP Upgrade: IP: 192.168.51.1, Server: 192.168.51.100

MAC addresses as verified by OEM firmware:

use   address   source
2g    *:d8      factory 0x0004 (label)
5g    *:d9      factory 0x8004
LAN   *:d7      factory $label -1
WAN   *:da      factory $label +2

Installation (TFTP + U-Boot):
 * Connect device with a TTL cable and open a serial session by PuTTY.
 * Press "2" when booting to select "Load system code then write to Flash via TFTP".
 * Configure the IP of local host server.
 * Upload firmware by tftpd64, it will boot when write instruction is executed.

Signed-off-by: Shiji Yang <yangshiji66@qq.com>

DragonBluep commented:
> Looks good so far.
> 
> MAC addresses are a bit unexpected (with -1 and +2 for the offsets). Are there addresses in 0x28, 0x2e, 0xe000, 0xe006?
> 
> Please also add an overview to the commit message:
> https://openwrt.org/docs/guide-developer/device-support-policies#mac_addresses
> 
> Any chance to have a flashing method without serial access?

MAC addresses as verified by OEM firmware:

use   address   source
2g    *:d8      factory 0x0004 (label)
5g    *:d9      factory 0x8004
LAN   *:d7      factory $label -1
WAN   *:da      factory $label +2

From factory eeprom layout, the WLAN 5g mac is $label +1 so it's not strange that WAN can only use $label +2. And there are just “0xff” value in factory@0x2e and 0x28.

I can remember it seems like this device can't apply update from local file so I did not find a good way to flash image except tftp.

DragonBluep commented:
![2021-08-24 225228](https://user-images.githubusercontent.com/70847398/130646121-bd8ed82e-917a-45de-8794-df946be382ed.jpg)
![2021-08-24 225441](https://user-images.githubusercontent.com/70847398/130646127-8fda9645-2e13-4ee0-8275-f67fc7af7bec.jpg)

adschm commented:
That means that "label-mac-device = &ethernet" is wrong and we need to set label-mac-address via 02_network ...?

adschm commented:
I've now implemented the label MAC address change myself and will merge this later today.

Thanks.
